### DISCUSSION TITLE
OffScreenRenderer with Python and Docker on Linux

### DISCUSSION THREAD

**Discussion by dancergraham** [isl-org/Open3D#5001] [Q&A] [answered]
Hello,

I am trying to run the OffScreenRenderer on a remote Linux server under Docker.  I am not very comfortable with Linux or Docker! I have tried using the [Docker example from Open3d](http://www.open3d.org/docs/0.10.0/tutorial/docker/index.html) copied below with the [image rendering example file](https://raw.githubusercontent.com/isl-org/Open3D/master/examples/python/visualization/render_to_image.py).

Any help or advice is very welcome

I use `docker build --tag open3d_test . ` and then I have tried various ways to run it with no success:

## No switches 

`docker run open3d_test` 

```bash
FEngine (64 bits) created at 0x7fe6a5b93010 (threading is enabled)
FEngine resolved backend: OpenGL
Failed to open X display. (exiting).
```

## Intel GPU switches 

`docker run --device=/dev/dri:/dev/dri o3dtest` 

```bash
docker: Error response from daemon: error gathering device information while adding custom device "/dev/dri": no such file or directory.
ERRO[0000] error waiting for container: context canceled
```

## X Server switches 

`docker run -v /tmp/.X11-unix:/tmp/.X11-unix -e DISPLAY o3dtest` 

```bash
FEngine (64 bits) created at 0x7f77d9852010 (threading is enabled)
FEngine resolved backend: OpenGL
Failed to open X display. (exiting).
```

## Dockerfile

```Dockerfile
# This could also be another Ubuntu or Debian based distribution
FROM python:3.9-bullseye

# Install Open3D system dependencies and pip
RUN apt-get update && apt-get install --no-install-recommends -y \
    libgl1 \
    libgomp1 \
    python3-pip \
    && rm -rf /var/lib/apt/lists/*

# Install Open3D from the pypi repositories
RUN python3 -m pip install --no-cache-dir --upgrade open3d
# Copy python file into the docker container and run it
COPY render_to_image.py .
CMD python3 -u render_to_image.py
```

**Comment by dancergraham** [comment, marked as answer]
** UPDATE ; See reply  below for solution **

The following Dockerfile, using `xvfb` with `xvfb-run` and requiring `xauth` seems to work, in that it generates the test images but then something clearly goes very wrong afterwards.  I am able to work around this by running the python code via `multiprocessing` as described here: https://github.com/isl-org/Open3D/discussions/4437


### Output

```shell
FEngine (64 bits) created at 0x7f167660f010 (threading is enabled)
FEngine resolved backend: OpenGL
Saving image at test.png
Saving image at test2.png
corrupted size vs. prev_size in fastbins
Aborted (core dumped)
```

### Dockerfile

```Dockerfile
# This could also be another Ubuntu or Debian based distribution
FROM python:3.9-bullseye

# Install Open3D system dependencies and pip
RUN apt-get update && apt-get install --no-install-recommends -y \
    libgl1 \
    xauth \
    xvfb \
    libgomp1 \
    python3-pip \
    && rm -rf /var/lib/apt/lists/*

# Install Open3D from the pypi repositories
RUN python3 -m pip install --no-cache-dir --upgrade open3d
COPY render_to_image.py .
CMD xvfb-run -a python3 -u render_to_image.py
```

**Reply by forrestjgq** [reply]
1. make sure remote server has X11 supported.
2. make sure docker container has correct $DISPLAY env
3. If you're using NVidia graphic device, I'll suggest you to run docker by nvidia-docker
4. If you run it from local ssh client, make sure ssh X11 forwarding is enabled.

**Reply by dancergraham** [reply]
Great - thank you very much
